### ISSUE TITLE
Field Regularization Suggestions

### ISSUE THREAD

**Issue by jiadingfang** [pfnet-research/distilled-feature-fields#5]
Thanks for the great work and the code!

I'm trying to implement DFF in my own codebase. However, what I noticed is that, while the feature loss is getting down and 2D rendered feature possess good quality, the 3D feature field itself can be very noisy, i.e., it only has continuity around a very tiny region of the query point, which can be undesirable in many cases. I wonder if there are any regularizations you applied to help alleviate the  problem.

I noticed you mentioned `stop-gradient` in the paper when training the feature field to avoid the problem that SSL features can be multi-view inconsistent. But I'm not exactly what does that mean in implementation, or do you have a reference that I can look for in this codebase?

I would also like to hear about any other suggestions that you might have in your experiments. 

Once again, thanks for your work and I would highly appreciate it if you could respond!

**Comment by soskek** [comment]
Thank you for comments!

>it only has continuity around a very tiny region of the query point, which can be undesirable in many cases.

Could you describe any case? I can imagine some. But, if described in detail, I could give better comments.

>I wonder if there are any regularizations you applied to help alleviate the problem.

Typical solutions would be
- Learning rate schedule. The early learning stage of NeRF provides some density value to wide regions loosely. So, optimizing feature fields more in the early stage may lead to a different tendency of optimization.
- Tuning the frequency bias of the feature MLP. As we discussed in the paper, we often expect a feature MLP to work with lower frequency properties than the radiance field MLP. So, for example, it could be a nice trick to prepare an independent feature MLP with low-frequency positional encoding.
- Increasing the training view. It also increases the training feature maps from various views and improves optimization as 3D space.
- And, maybe the point below is also related to the problem. See it too.

>I noticed you mentioned stop-gradient in the paper when training the feature field to avoid the problem that SSL features can be multi-view inconsistent. 

Actually, I just discussed stop-gradient to avoid noising the training of the density field. Because feature map supervisions are multi-view inconsistent, if the training loss backprops to the density field, it could be a noise, which makes NeRF itself worse.

First, I'm sorry that this part is not fully implemented in this NGP-based codebase to keep the simplicity of the cuda code and make it much more memory efficient in the NGP framework. This repository is a specific variant of the DFF model and has some differences from the paper's one.
Rendering a feature via volume rendering only uses a tiny set of samples (5 points) around the rendered surface. The code is [here](https://github.com/pfnet-research/distilled-feature-fields/blob/master/models/rendering.py#L188-L192). This is simple but may lead to worse results sometimes.

Secondly, anyway, I can show a simple TRUE reference code.
When the volume rendering of RGB is simply written as
```
# xyzs.shape == (N_rays, M_samples, 3); each sample's coordinate
# dirs.shape == (N_rays, M_samples, 3); each sample's view direction
# ts.shape == (N_rays, M_samples); each sample's depth value
densities = calc_density(xyzs)  # (N_rays, M_samples)
weights = calc_weights(densities, ts)  # (N_rays, M_samples)
rgbs = calc_rgb(xyzs, dirs)  # (N_rays, M_samples, 3)
rgb = (rgbs * weights[:, :, None]).sum(1)  # (N_rays, 3)
```
we can write its feature rendering version as follows
```
features = calc_feature(xyzs)  # (N_rays, M_samples, D)
feature = (features * weights[:, :, None].detach()).sum(1)  # (N_rays, D)
```
Very simple. We just call `.detach()` for weights in the weighted sum.
